File: source/rc2/Rc2RestServer.m
//
//  Rc2RestServer.m
//  Rc2Client
//
//  Created by Mark Lilback on 12/10/15.
//  Copyright © 2015 West Virginia University. All rights reserved.
//

#import "Rc2RestServer.h"
#import "Rc2AppConstants.h"
#import "NSArray+Rc2Extensions.h"
#import "Rc2-Swift.h"

static Rc2RestServer *sInstance;
static NSString *const kServerHostKey = @"ServerHostKey";

///posted on login and logout with the object being the Rc2RestServer instance
NSString * const Rc2RestLoginStatusChangedNotification = @"Rc2RestLoginStatusChangedNotification";


@interface Rc2RestServer ()
@property (nonatomic, strong) NSURLSessionConfiguration *urlConfig;
@property (nonatomic, strong, readwrite) NSURLSession *urlSession;
@property (nonatomic, strong, readwrite) Rc2LoginSession *loginSession;
@property (nonatomic, copy) NSArray *hosts;
@property (nonatomic, copy) NSURL *baseUrl;
@end

@implementation Rc2RestServer
+(instancetype)sharedInstance {
	if (nil == sInstance) {
		static dispatch_once_t onceToken;
		dispatch_once(&onceToken, ^{
			sInstance = [[Rc2RestServer alloc] init];
		});
	}
	return sInstance;
}

+(void)setSharedInstance:(Rc2RestServer*)server
{
	sInstance = server;
}

-(id)initWithSessionConfiguration:(NSURLSessionConfiguration*)config
{
	if ((self = [super init])) {
		_urlConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
		_urlConfig.HTTPAdditionalHeaders = @{@"User-Agent": [self userAgentString], @"Accept":@"application/json"};
		_urlSession = [NSURLSession sessionWithConfiguration:_urlConfig];
		NSError *err;
		NSURL *hostFileUrl = [[NSBundle mainBundle] URLForResource:@"Rc2RestHosts" withExtension:@"json"];
		NSAssert(hostFileUrl != nil, @"failed to read Rc2RestHosts.json");
		NSData *jsonData = [NSData dataWithContentsOfURL:hostFileUrl];
		NSAssert(jsonData, @"Failed to read json data from Rc2RestHosts file");
		NSDictionary *json = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&err];
		self.hosts = [json objectForKey:@"hosts"];
		NSAssert(self.restHosts.count > 0, @"invalid hosts data");
	}
	return self;
}

-(id)init
{
	return [self initWithSessionConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
}

#pragma mark - basic public methods

-(NSArray<NSString*>*)restHosts
{
	return [self.hosts valueForKeyPath:@"name"];
}

-(NSString*)defaultRestHost
{
	return [[self.hosts objectAtIndex:[[NSUserDefaults standardUserDefaults] integerForKey:kServerHostKey]] valueForKey:@"name"];
}

-(NSString*)userAgentString
{
#if (__MAC_OS_X_VERSION_MIN_REQUIRED >= 1090)
	return @"Rc2 MacClient";
#else
	return @"Rc2 iPadClient";
#endif
}

-(NSString*)connectionDescription
{
	NSString *login = [self.loginSession valueForKeyPath:@"currentUser.login"];
	NSString *host = [self.loginSession valueForKey:@"host"];
	if ([host isEqualToString:@"rc2"])
		return login;
	return [NSString stringWithFormat:@"%@@%@", login, host];
}

#pragma mark - internal utility methods

-(NSMutableURLRequest*)requestWithPath:(NSString*)path method:(NSString*)method json:(NSDictionary*)jsonDict
{
	NSError *error;
	NSURL *url = [NSURL URLWithString:path relativeToURL:self.baseUrl];
	NSMutableURLRequest *req = [NSMutableURLRequest requestWithURL:url];
	req.HTTPMethod = method;
	if (self.loginSession.authToken)
		[req addValue:self.loginSession.authToken forHTTPHeaderField:@"Rc2-Auth"];
	if (jsonDict.count > 0) {
		[req addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
		req.HTTPBody = [NSJSONSerialization dataWithJSONObject:jsonDict options:0 error:&error];
		if (error)
			Rc2LogError(@"error serialzing to json: %@ (%@)", jsonDict, error.localizedDescription);
	}
	return req;
}

#pragma mark - login/logout

-(void)loginToHostName:(NSString*)hostName login:(NSString*)login password:(NSString*)password handler:(Rc2RestCompletionHandler)handler
{
	//get the host structure
	NSDictionary *hostDict = [self.hosts rc2_firstObjectWithValue:hostName forKey:@"host"];
	if (nil == hostDict) {
		[NSException raise:NSInvalidArgumentException format:@"invalid hostname: %@", hostName];
	}
	NSString *hoststr = [NSString stringWithFormat:@"http%@://%@:%@/",
						 [hostDict[@"secure"] boolValue] ? @"s" : @"",
						 hostDict[@"host"],
						 hostDict[@"port"]];
	self.baseUrl = [NSURL URLWithString:hoststr];
	//setup done, make login request
	NSMutableURLRequest *req = [self requestWithPath:@"login" method:@"POST" json:@{@"login":login, @"password":password}];
	NSURLSessionDataTask *task = [self.urlSession dataTaskWithRequest:req completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error)
	{
		NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
		NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
		if (httpResponse.statusCode == 401) {
			NSError *loginError = [NSError errorWithDomain:Rc2ErrorDomain code:401 userInfo:@{NSLocalizedDescriptionKey:NSLocalizedString(@"Invalid login or password", @"")}];
			dispatchOnMainQueue( ^{handler(NO, nil, loginError); } );
		} else if (httpResponse.statusCode == 200) {
			self.loginSession = [[Rc2LoginSession alloc] initWithJsonData:json host:hostDict[@"host"]];
			dispatchOnMainQueue( ^{handler(YES, self.loginSession, nil);} );
			NSInteger hostIndex = [self.hosts indexOfObject:hostDict];
			[[NSUserDefaults standardUserDefaults] setInteger:hostIndex forKey:kServerHostKey];
			[[NSNotificationCenter defaultCenter] postNotificationName:Rc2RestLoginStatusChangedNotification object:self];
		} else {
			Rc2LogWarn(@"login got unknown error:%ld", (long)httpResponse.statusCode);
			dispatchOnMainQueue( ^{handler(NO, nil, error); });
		}
	}];
	[task resume];
}

#pragma mark - workspaces

//updates the workspaces array of the loginSession
-(void)createWorkspace:(NSString*)wspaceName completionBlock:(Rc2RestCompletionHandler)handler
{
	NSMutableURLRequest *req = [self requestWithPath:@"workspaces" method:@"POST" json:@{@"name":wspaceName}];
	NSURLSessionDataTask *task = [self.urlSession dataTaskWithRequest:req completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error)
	{
		NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
		NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
		if (httpResponse.statusCode == 200) {
			Rc2Workspace *wspace = [[Rc2Workspace alloc] initWithJsonData:json];
			NSMutableArray *spaces = [self.loginSession.workspaces mutableCopy];
			[spaces addObject:wspace];
			[spaces sortUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES selector:@selector(localizedCaseInsensitiveCompare:)]]];
			self.loginSession.workspaces = spaces;
			dispatchOnMainQueue( ^{handler(YES, wspace, nil);} );
		} else {
			Rc2LogWarn(@"create workspace got unknown error:%ld", (long)httpResponse.statusCode);
			dispatchOnMainQueue( ^{handler(NO, nil, error); });
		}
	}];
	[task resume];
}

-(void)renameWorkspce:(Rc2Workspace*)wspace name:(NSString*)newName completionHandler:(Rc2RestCompletionHandler)handler
{
	NSString *path = [NSString stringWithFormat:@"workspaces/%d", wspace.wspaceId];
	NSMutableURLRequest *req = [self requestWithPath:path method:@"PUT" json:@{@"name":newName, @"id":@(wspace.wspaceId)}];
	NSURLSessionDataTask *task = [self.urlSession dataTaskWithRequest:req completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error)
	{
		NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
		NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
		if (httpResponse.statusCode == 200) {
			Rc2Workspace *modSpace = [[Rc2Workspace alloc] initWithJsonData:json];
			NSMutableArray *spaces = [self.loginSession.workspaces mutableCopy];
			[spaces replaceObjectAtIndex:[spaces indexOfObject:wspace] withObject:modSpace];
			[spaces sortUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES selector:@selector(localizedCaseInsensitiveCompare:)]]];
			self.loginSession.workspaces = spaces;
			dispatchOnMainQueue( ^{handler(YES, modSpace, nil);} );
		} else {
			Rc2LogWarn(@"rename workspace got unknown error:%ld", (long)httpResponse.statusCode);
			dispatchOnMainQueue( ^{handler(NO, nil, error); });
		}
	}];
	[task resume];
}

-(void)deleteWorkspce:(Rc2Workspace*)wspace completionHandler:(Rc2RestCompletionHandler)handler
{
	NSString *path = [NSString stringWithFormat:@"workspaces/%d", wspace.wspaceId];
	NSMutableURLRequest *req = [self requestWithPath:path method:@"DELETE" json:nil];
	NSURLSessionDataTask *task = [self.urlSession dataTaskWithRequest:req completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error)
	{
		NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
		NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
		if (httpResponse.statusCode == 200) {
			NSMutableArray *spaces = [self.loginSession.workspaces mutableCopy];
			[spaces removeObject:wspace];
			self.loginSession.workspaces = spaces;
			dispatchOnMainQueue( ^{handler(YES, nil, nil);} );
		} else {
			Rc2LogWarn(@"delete workspace got unknown error:%ld", (long)httpResponse.statusCode);
			dispatchOnMainQueue( ^{handler(NO, nil, error); });
		}
	}];
	[task resume];
}

@end

File: source/rc2/Rc2RestServer.swift
//
//  Rc2RestServer.swift
//  Rc2Client
//
//  Created by Mark Lilback on 12/15/15.
//  Copyright © 2015 West Virginia University. All rights reserved.
//

import Foundation


@objc public class Rc2RestServer : NSObject {

	private static var sInstance = Rc2RestServer()
	private let kServerHostKey = "ServerHostKey"
	
	///singleton accessor
	public static var sharedInstance : Rc2RestServer {
		get {
			return sInstance
		}
	}
	
	public typealias Rc2RestCompletionHandler = (success:Bool, results:AnyObject?, error:NSError?) -> Void
	
	private var urlConfig : NSURLSessionConfiguration
	private var urlSession : NSURLSession
	private(set) public var hosts : [NSDictionary]
	private(set) public var selectedHost : NSDictionary
	private(set) public var loginSession : Rc2LoginSession?
	private var baseUrl : NSURL

	var restHosts : [String] {
		get {
			return hosts.map({ $0["name"]! as! String })
		}
	}
	var connectionDescription : String {
		get {
			let login = loginSession?.currentUser.login
			let host = loginSession?.host
			if (host == "rc2") {
				return login!;
			}
			return "\(login)@\(host)"
		}
	}
	
	//private init so instance is unique
	override init() {
		var userAgent = "Rc2 iOSClient"
		#if os(OSX)
			userAgent = "Rc2 MacClient"
		#endif
		urlConfig = NSURLSessionConfiguration.defaultSessionConfiguration()
		urlConfig.HTTPAdditionalHeaders = ["User-Agent": userAgent, "Accept": "application/json"]
		urlSession = NSURLSession.init(configuration: urlConfig)
		
		//load hosts info from resource file
		let hostFileUrl = NSBundle.mainBundle().URLForResource("Rc2RestHosts", withExtension: "json")
		assert(hostFileUrl != nil, "failed to get Rc2RestHosts.json URL")
		let jsonData = NSData(contentsOfURL: hostFileUrl!)
		let json = JSON(data:jsonData!)
		let theHosts = json["hosts"].arrayObject!
		assert(theHosts.count > 0, "invalid hosts data")
		hosts = theHosts as! [NSDictionary]
		selectedHost = hosts.first!
		self.baseUrl = NSURL() //dummy place holder
		super.init()
		if let previousHostName = NSUserDefaults.standardUserDefaults().stringForKey(self.kServerHostKey) {
			selectHost(previousHostName)
		}
	}
	
	private func createError(code:Int, description:String) -> NSError {
		return NSError(domain: Rc2ErrorDomain, code: code, userInfo: [NSLocalizedDescriptionKey:NSLocalizedString(description, comment: "")])
	}
	
	private func request(path:String, method:String, jsonDict:NSDictionary) -> NSMutableURLRequest {
		let url = NSURL(string: path, relativeToURL: baseUrl)
		let request = NSMutableURLRequest(URL: url!)
		request.HTTPMethod = method
		if loginSession != nil {
			request.addValue(loginSession!.authToken, forHTTPHeaderField:"Rc-2Auth")
		}
		if (jsonDict.count > 0) {
			request.addValue("application/json", forHTTPHeaderField:"Content-Type")
			request.HTTPBody = try! NSJSONSerialization.dataWithJSONObject(jsonDict, options: [])
		}
		return request
	}
	
	public func selectHost(hostName:String) {
		if let hostDict = hosts.filter({ return ($0["name"] as! String) == hostName }).first {
			selectedHost = hostDict
			let hprotocol = hostDict["secure"]!.boolValue! ? "https" : "http"
			let hoststr = "\(hprotocol)://\(hostDict["host"]!.stringValue):\(hostDict["port"]!.stringValue)/"
			baseUrl = NSURL(string: hoststr)!
		}
	}
	
	public func login(login:String, password:String, handler:Rc2RestCompletionHandler) {
		let req = request("login", method:"POST", jsonDict: ["login":login, "password":password])
		let task = urlSession.dataTaskWithRequest(req) {
			(data, response, error) -> Void in
			let json = JSON(data:data!)
			let httpResponse = response as! NSHTTPURLResponse
			switch(httpResponse.statusCode) {
				case 200:
					self.loginSession = Rc2LoginSession(json: json, host: self.selectedHost["name"]! as! String)
					dispatch_async(dispatch_get_main_queue(), { handler(success: true, results: self.loginSession!, error: nil) })
					NSUserDefaults.standardUserDefaults().setObject(self.loginSession!.host, forKey: self.kServerHostKey)
					NSNotificationCenter.defaultCenter().postNotificationName(Rc2RestLoginChangedNotification, object: self)
				case 401:
					let error = self.createError(401, description: "Invalid login or password")
					Rc2LogVerbose("\(__FUNCTION__) got a \(httpResponse.statusCode)")
					dispatch_async(dispatch_get_main_queue(), { handler(success: false, results: nil, error: error) })
				default:
					let error = self.createError(httpResponse.statusCode, description: "")
					Rc2LogWarn("\(__FUNCTION__) got unknown error: \(httpResponse.statusCode)")
					dispatch_async(dispatch_get_main_queue(), { handler(success: false, results: nil, error: error) })
			}
		}
		task.resume()
	}
	
	public func createWorkspace(wspaceName:String, handler:Rc2RestCompletionHandler) {
		let req = request("workspaces", method:"POST", jsonDict: ["name":wspaceName])
		let task = urlSession.dataTaskWithRequest(req) { (data, response, error) -> Void in
			let json = JSON(data:data!)
			let httpResponse = response as! NSHTTPURLResponse
			switch(httpResponse.statusCode) {
			case 200:
				let wspace = Rc2Workspace(json: json)
				var spaces = (self.loginSession?.workspaces)!
				spaces.append(wspace)
				self.loginSession?.workspaces = spaces
				dispatch_async(dispatch_get_main_queue(), { handler(success: true, results: wspace, error: nil) })
			case 422:
				let error = self.createError(422, description: "A workspace with that name already exists")
				Rc2LogVerbose("\(__FUNCTION__) got duplicate error")
				dispatch_async(dispatch_get_main_queue(), { handler(success: false, results: nil, error: error) })
			default:
				let error = self.createError(httpResponse.statusCode, description: "")
				Rc2LogWarn("\(__FUNCTION__) got unknown error: \(httpResponse.statusCode)")
				dispatch_async(dispatch_get_main_queue(), { handler(success: false, results: nil, error: error) })
			}
		}
		task.resume()
	}

	public func renameWorkspace(wspace:Rc2Workspace, newName:String, handler:Rc2RestCompletionHandler) {
		let req = request("workspaces/\(wspace.wspaceId)", method:"PUT", jsonDict: ["name":newName, "id":Int(wspace.wspaceId)])
		let task = urlSession.dataTaskWithRequest(req) { (data, response, error) -> Void in
			let json = JSON(data!)
			let httpResponse = response as! NSHTTPURLResponse
			switch(httpResponse.statusCode) {
			case 200:
				let modWspace = Rc2Workspace(json: json)
				var spaces = (self.loginSession?.workspaces)!
				spaces[spaces.indexOf(wspace)!] = modWspace
				dispatch_async(dispatch_get_main_queue(), { handler(success: true, results: modWspace, error: nil) })
			default:
				let error = self.createError(httpResponse.statusCode, description: "")
				Rc2LogWarn("\(__FUNCTION__) got unknown error: \(httpResponse.statusCode)")
				dispatch_async(dispatch_get_main_queue(), { handler(success: false, results: nil, error: error) })
			}
		}
		task.resume()
	}

	
	public func deleteWorkspace(wspace:Rc2Workspace, handler:Rc2RestCompletionHandler) {
		let req = request("workspaces/\(wspace.wspaceId)", method:"DELETE", jsonDict: [:])
		let task = urlSession.dataTaskWithRequest(req) { (data, response, error) -> Void in
			let httpResponse = response as! NSHTTPURLResponse
			switch(httpResponse.statusCode) {
			case 204:
				self.loginSession?.workspaces.removeAtIndex((self.loginSession?.workspaces.indexOf(wspace))!)
				dispatch_async(dispatch_get_main_queue(), { handler(success: true, results: nil, error: nil) })
			default:
				let error = self.createError(httpResponse.statusCode, description: "")
				Rc2LogWarn("\(__FUNCTION__) got unknown error: \(httpResponse.statusCode)")
				dispatch_async(dispatch_get_main_queue(), { handler(success: false, results: nil, error: error) })
			}
		}
		task.resume()
	}

}

